src/mzcloud-auth/src/lib.rs

```diff
@@ -0,0 +1,115 @@
+pub struct MzCloudAuthentication {
+    pub admin_api_token_url: String,
+    #[derivative(Debug = "ignore")]
+    pub decoding_key: DecodingKey,
+    pub tenant_id: String,
+}
+
+impl MzCloudAuthentication {
+    pub fn new(
+        admin_api_token_url: String,
+        rsa_pem: &[u8],
+        tenant_id: String,
+    ) -> Result<Self, anyhow::Error> {
+        let decoding_key = DecodingKey::from_rsa_pem(rsa_pem)?;
+        Ok(Self {
+            admin_api_token_url,
+            decoding_key,
+            tenant_id,
+        })
+    }
+
+    // Exchanges a password (clientid and secret) for a jwt token and validates
+    // it. Password must have the format: "clientid,secret".
+    pub async fn validate_password(&self, password: &str) -> Result<Claims, anyhow::Error> {
+        let token = self.exchange_password_for_token(password).await?;
+        self.validate_access_token(&token.access_token)
+    }
+
+    // Exchanges a password (clientid and secret) for a jwt token. Password must
+    // have the format: "clientid,secret".
+    pub async fn exchange_password_for_token(
+        &self,
+        password: &str,
+    ) -> Result<ApiTokenResponse, anyhow::Error> {
+        let (client_id, secret) = match password.split_once(',') {
+            Some(sp) => sp,
+            None => anyhow::bail!("expected comma-delimiter client+secret"),
+        };
+        self.exchange_client_secret_for_token(client_id, secret)
+            .await
+    }
+
+    // Exchanges a client id and secret for a jwt token.
+    pub async fn exchange_client_secret_for_token(
+        &self,
+        client_id: &str,
+        secret: &str,
+    ) -> Result<ApiTokenResponse, anyhow::Error> {
+        let resp = reqwest::Client::new()
+            .post(&self.admin_api_token_url)
+            .json(&ApiTokenArgs { client_id, secret })
+            .send()
+            .await?
+            .error_for_status()?
+            .json::<ApiTokenResponse>()
+            .await?;
+
+        Ok(resp)
+    }
+
+    // Validates an access token and its `tenant_id`.
+    pub fn validate_access_token(&self, token: &str) -> Result<Claims, anyhow::Error> {
+        let msg = decode::<Claims>(
+            &token,
+            &self.decoding_key,
+            &Validation::new(Algorithm::RS256),
+        )?;
+        if msg.claims.tenant_id != self.tenant_id {
+            anyhow::bail!("tenant ids don't match")
+        }
+        Ok(msg.claims)
+    }
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ApiTokenArgs<'a> {
+    pub client_id: &'a str,
+    pub secret: &'a str,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ApiTokenResponse {
+    pub expires: String,
+    pub expires_in: i64,
+    pub access_token: String,
+    pub refresh_token: String,
+}
+
+// TODO: Do we care about the sub? Do we need to validate the sub or other
+// things, even if unused?
```

Comment:
That reminds me: we should double check in our shared Slack channel with Frontegg if they consider email to be a stable identifier.

Reply:
Can you define stable here? Does it mean that frontegg doesn't allow creating multiple users with the same email? That the email is unchangeable? What about if a user is deleted then recreated with the same email, does that get a new sub (do we care? maybe we don't care about the sub?).

Does it matter if the email address is not whatever stable means? What problems occur if an email changes? mz would get the new email on next login and auto create that user, and any permissions (after we add those) would apply to that address. I suppose one problem could be that if the clientid,secret hasn't changed but the email address did, then there could be two sessions existing at the same time with different usernames but equal passwords, which certainly seems wrong. Ok so maybe the specific thing we would like to know is if a clientid will always return the same email address.